### PR TITLE
feat(SwapsController): Remove reliance on global network

### LINKED ISSUE #23566: Make SwapsController support Amon Hen v2
### What is this about?

Make SwapsController support Amon Hen v2

### Scenario

_No response_

### Design

_No response_

### Technical Details

Make SwapsController support Amon Hen v2

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [X] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [ ] Other (please specify)

### References

_No response_

### PR DESCRIPTION
**Pull request by mcmire**
## **Description**

Currently, SwapsController relies on the global network: it uses the global provider from the NetworkController and feeds it into an Ethers contract object to get the allowance for a user's token, and it uses the chain ID of the global network (again, coming from NetworkController) to retrieve parameters that will ultimately be used to control when and how swaps are requested and how to transform the resulting data.

This does not work long-term, as we want to remove the concept of the global network from the extension. So fundamentally, we need to have the parts of the API which rely on the global provider to take a network client ID, which will be used to fetch a provider and chain ID.

One important note about SwapsController is that there is a kind of entry point to the API via `fetchAndSetQuotes`. This establishes the network that a polling loop will use, and it starts that loop. The controller also operates on one network at a time: it does not distinguish between one network or another when capturing the quote data in state, so when calling `fetchAndSetQuotes` for a different network, if the polling loop was already started for a different network, it will be reset, and all quotes captured for that network will be overwritten. We plan on keeping this behavior.

Given these constraints, here is the list of changes:

- Update the `FetchTradesInfoParams` type to replace `chainId` with `networkClientId`.
- Update `fetchAndSetQuotes` to take a `networkClientId` instead of `chainId` within the `fetchParamsMetaData` (second argument). When called for the first time, this ID will be resolved to a network client.
- Update `fetchAndSetQuotes` to no longer rely on the global provider proxy from NetworkController to get the allowance for a token, but use the network client obtained in the previous step.
- Update signature for `getTopQuoteWithCalculatedSavings`: instead of taking a single quotes object, it now takes an options bag with keys `quotes` and `networkClientId`. Additionally, instead of relying on the global network to get the chain ID, it will resolve the network client ID to a network client and get the chain ID off of that.
- Add new selector `getSelectedNetwork` which can be used to get information about the global network configuration + network client ID in a Redux action in one go (`getCurrentNetwork` was not sufficient to do this).

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/28275?quickstart=1)

## **Related issues**

Progresses #23566.

## **Manual testing steps**

1. Switch to Ethereum Mainnet if you haven't already.
2. Click on the "Swap" button.
3. Open Dev Tools and go to the Network tab.
4. Fill in a monetary value for ETH.
5. Choose a destination token (it doesn't matter).
6. Wait for quotes to appear.
7. Observe that the text above the quotes says "Confirmed by X sources. Verify on Etherscan."
8. In the Dev Tools pane, search for "swap.api". Find any request that starts with `https://swap.api.cx.metamask.io/networks/`. Observe that the chain ID in the URL matches Ethereum Mainnet.
9. Now search for "gas.api". Find any request that starts with `https://gas.api.cx.metamask.io/networks/`. Observe that the chain ID in the URL matches Ethereum Mainnet.
10. Cancel the swap.
11. Switch to Linea Mainnet.
12. Fill in a monetary value for ETH.
13. Choose a destination token.
14. Wait for quotes to appear.
15. Observe that the text above the quotes says "Confirmed by X sources. Verify on LineaScan."
16. In the Dev Tools pane, search for "swap.api". Find any request that starts with `https://swap.api.cx.metamask.io/networks/`. Observe that the chain ID in the URL matches Linea Mainnet.
17. Now search for "gas.api". Find any request that starts with `https://gas.api.cx.metamask.io/networks/`. Observe that the chain ID in the URL matches Linea Mainnet.

## **Screenshots/Recordings**

No screenshots/recordings necessary. Everything should work exactly the same as it does today.

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/selectors/selectors.js [hunk 1]
```diff
@@ -1916,6 +1916,35 @@ export const getCurrentNetwork = createDeepEqualSelector(
   },
 );
 
+export const getSelectedNetwork = createDeepEqualSelector(
```
Comment (mcmire) [inline ui/selectors/selectors.js:1919]: I realize that we already have `getProviderConfig`, but unfortunately that doesn't give me what I need. I need both the `selectedNetworkClientId` and the chain ID that corresponds to the selected network. I guess I could have reused `getCurrentChainId` in this case, but I'm already making use of `networkConfigurationsByChainId` and it gives me everything I need.

File: yarn.lock [hunk 2]
```diff
@@ -5212,6 +5212,19 @@ __metadata:
   languageName: node
   linkType: hard
 
+"@metamask/eth-json-rpc-provider@npm:^4.1.6":
```
[the remark this answers is not in the document]
Comment (mcmire) [reply yarn.lock]: Makes sense!

**Comment by mcmire** [comment]
@metamaskbot update-policies

**Comment by mcmire** [comment]
I've resolved the conflicts. This PR is ready for review again.

**Review by mikesposito — COMMENTED** [review]
Looks great! 
Only one question before approving: do we have a way to test the controller on test networks?

**Review by mikesposito — DISMISSED** [review]
Tested with a Swap on OP Mainnet, LGTM!

https://github.com/user-attachments/assets/3997732c-76cc-4543-8901-a3cb3bb0558f

**Comment by mcmire** [comment]
@mikesposito Oh super, thank you!

**Review by mikesposito — APPROVED** [review]

File: ui/selectors/selectors.js [hunk 3]
```diff
@@ -2012,6 +2012,35 @@ export const getCurrentNetwork = createDeepEqualSelector(
...
+
+    // TODO: Add `networkConfigurationsByNetworkClientId` to NetworkController state so this is easier to do
+    const possibleNetworkConfiguration = Object.values(
+      networkConfigurationsByChainId,
+    ).find((networkConfiguration) => {
+      return networkConfiguration.rpcEndpoints.some((rpcEndpoint) => {
+        return rpcEndpoint.networkClientId === selectedNetworkClientId;
+      });
+    });
+    if (possibleNetworkConfiguration === undefined) {
```
Comment (infiniteflower) [inline ui/selectors/selectors.js:2031]: Does this case happen often? Or is this pretty edge case?

**Comment by infiniteflower** [comment]
Tested and did a swap on Linea, working fine

**Review by infiniteflower — APPROVED** [review]

**Pull request merged**